Tests/Realm.Tests/Database/CollectionTests.cs

```diff
@@ -411,6 +414,258 @@ public void Results_GetFiltered_SanityTest()
             Assert.That(query.ToArray().All(i => i.Int >= 5));
         }
 
+        public struct StringQueryTestData
+        {
+            public string PropertyName;
+            public RealmValue MatchingValue;
+            public RealmValue NonMatchingValue;
+
+            private delegate object Converter(RealmValue value);
+
+            public StringQueryTestData(string propertyName, RealmValue matchingValue, RealmValue nonMatchingValue)
+            {
+                PropertyName = propertyName;
+                MatchingValue = matchingValue;
+                NonMatchingValue = nonMatchingValue;
+            }
+
+            public (object? MatchingValue, object? NonMatchingValue) ConvertBoxedValueToCorrectType(Type targetType)
+            {
+                if (targetType == null)
+                {
+                    return (null, null);
+                }
+
+                // boxed value types can't be casted to Decimal128
+                if (PropertyName == "Decimal128Property")
+                {
+                    return (TryToConvert(MatchingValue, (value) => value.AsDecimal128()), TryToConvert(NonMatchingValue, (value) => value.AsDecimal128()));
+                }
+                else
+                {
+                    return (TryToConvert(MatchingValue.AsAny(), targetType), TryToConvert(NonMatchingValue.AsAny(), targetType));
+                }
+            }
```

Comment:
Instead of this, let's try a simpler approach:

```csharp
public static object BoxValue(RealmValue val, Type targetType)
{
    var boxed = val.AsAny();
    if (boxed != null && boxed.GetType() != targetType)
    {
        boxed = Convert.ChangeType(boxed, targetType);
    }

    return boxed;
}

// And we use it as:

var boxedMatch = data.BoxValue(data.MatchingValue, prop.PropertyType);
var boxedNonMatch = data.BoxValue(data.NonMatchingValue, prop.PropertyType);
```